src/finn/custom_op/fpgadataflow/thresholding_binary_search.py

```diff
@@ -0,0 +1,752 @@
+        pe = self.get_nodeattr("PE")
+        n_thres_steps = self.get_nodeattr("numSteps")
+        decoupled_thres_pe_flipped = np.flip(decoupled_thres, axis=-2)
+        decoupled_thres = decoupled_thres.reshape(1, -1, pe * n_thres_steps)
+        decoupled_thres = decoupled_thres.copy()
+        decoupled_thres_pe_flipped = decoupled_thres_pe_flipped.reshape(
+            1, -1, pe * n_thres_steps
+        )
+        decoupled_thres_pe_flipped = decoupled_thres_pe_flipped.copy()
+
+        if weight_file_mode == "decoupled_npy":
+            # save weight stream into npy for cppsim
+            np.save(weight_file_name, decoupled_thres)
+        elif weight_file_mode == "decoupled_verilog_dat":
+            # convert weight values into hexstring
+            weight_width = self.get_weightstream_width()
+            # pad to nearest 4 bits to get hex strings
+            weight_width_padded = roundup_to_integer_multiple(weight_width, 4)
+            weight_tensor_pe_flipped = pack_innermost_dim_as_hex_string(
+                decoupled_thres_pe_flipped, tdt, weight_width_padded, prefix=""
+            )
+            weight_stream = weight_tensor_pe_flipped.flatten()
+            weight_stream = weight_stream.copy()
+            with open(weight_file_name, "w") as f:
+                for val in weight_stream:
+                    f.write(val + "\n")
+        elif weight_file_mode == "decoupled_runtime":
+            # memstream axi-lite interface will map each mem line to
+            # one or multiple 32-bit words
+            weight_width = self.get_weightstream_width()
+            words_per_memwidth = 2 ** ceil(log2(weight_width / 32))
+            if words_per_memwidth < 1:
+                words_per_memwidth = 1
+            weight_width_padded = words_per_memwidth * 32
+            # first, pack and ensure padding to 32 bits
+            weight_tensor_pe_flipped = pack_innermost_dim_as_hex_string(
+                decoupled_thres_pe_flipped, tdt, weight_width_padded, prefix=""
+            )
+            weight_stream = weight_tensor_pe_flipped.flatten()
+            weight_stream = weight_stream.copy()
+            with open(weight_file_name, "w") as f:
+                for val in weight_stream:
+                    # split into groups of 8 hex digits (= 32 bits)
+                    words_32b = textwrap.wrap(val, 8)
+                    words_32b.reverse()
+                    for word_32b in words_32b:
+                        f.write(word_32b + "\n")
+        else:
+            raise Exception("Decoupled weight export not yet implemented")
+
+    # Get the integer from the DataType and string-ify it
+    # This assumes that the data is in the form "INTx" or similar
+    def conv_datatype_to_str(self, data_type):
+        # Handle the case that an int is passed to the function
+        if isinstance(data_type, int):
+            return str(data_type)
+        return str(DataType[data_type].bitwidth())
```

Comment:
some code smells here:
- function name does not reflect what it does
- the parameter data_type' has unrestricted type, can be either a DataType string or an integer
- defined as class member function (taking self as argument) even though not accessing `self.`

these are not necessarily problems but flagging them anyway

Reply:
removed the function - it potentially causes more problems than it solves: https://github.com/Xilinx/finn/pull/687/commits/e663030e98dc6c1f194ccec1d8e5d65b9599c19c